File: BFS/네 방향 탈출 가능 여부 판별하기.py
from collections import deque

# 변수 선언 및 입력
n, m = map(int, input().split())
a = [list(map(int, input().split())) for _ in range(n)]

visited = [[False] * m for _ in range(n)]

# 주어진 위치가 격자를 벗어나는지 확인
def in_range(x, y):
    return 0 <= x < n and 0 <= y < m

# 주어진 위치로 이동 가능한지 확인
def can_go(x, y):
    return in_range(x, y) and a[x][y] and not visited[x][y]

# BFS 탐색
def bfs():
    q = deque([(0, 0)])  # 시작점 추가
    visited[0][0] = True
    dxs, dys = [0, 1, 0, -1], [1, 0, -1, 0]  # 방향 설정

    while q:
        x, y = q.popleft()
        for dx, dy in zip(dxs, dys):
            new_x, new_y = x + dx, y + dy
            if can_go(new_x, new_y):
                q.append((new_x, new_y))
                visited[new_x][new_y] = True

# BFS 수행
bfs()

# 우측 하단 방문 여부 확인
print(1 if visited[n - 1][m - 1] else 0)

File: BFS/BFS 탐색/네_방향_탈출_가능_여부_판별하기.java
import java.util.*;
import java.io.*;

class Pair {
    int x, y;

    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class Main {
    static final int MAX_N = 100;

    static int n, m;
    static int[][] arr = new int[MAX_N][MAX_N];
    static boolean[][] visited = new boolean[MAX_N][MAX_N];
    static Queue<Pair> queue = new LinkedList<>();

    static boolean inRange(int x, int y) {
        return 0 <= x && x < n && 0 <= y && y < m;
    }

    static boolean canGo(int x, int y) {
        return inRange(x, y) && arr[x][y] == 1 && !visited[x][y];
    }

    static void BFS() {
        while (!queue.isEmpty()) {
            Pair curr = queue.poll();
            int x = curr.x;
            int y = curr.y;

            int[] dx = new int[]{0, 1, 0, -1};
            int[] dy = new int[]{1, 0, -1, 0};

            for (int i = 0; i < 4; i++) {
                int newX = x + dx[i];
                int newY = y + dy[i];

                if (canGo(newX, newY)) {
                    visited[newX][newY] = true;
                    queue.add(new Pair(newX, newY));
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        queue.add(new Pair(0, 0));
        visited[0][0] = true;

        BFS();

        int answer = visited[n - 1][m - 1] ? 1 : 0;
        System.out.println(answer);
    }
}